doc/ref/chap-builtin-cmd.md

```diff
@@ -1186,3 +1186,188 @@ Flags:
+is intended to be used for command-line interfaces to YSH programs.
+
+To use it, first import `args.ysh`:
+
+    source --builtin args.ysh
+
+Then, create an argument parser **spec**ification:
+
+    parser (&spec) {
+      flag -v --verbose (help="Verbosely")  # default is Bool, false
+
+      flag -P --max-procs ('int', default=-1, help='''
+        Run at most P processes at a time
+        ''')
+
+      flag -i --invert ('bool', default=true, help='''
+        Long multiline
+        Description
+        ''')
+
+      arg src (help='Source')
+      arg dest (help='Dest')
+
+      rest files
+    }
+
+Finally, parse `ARGV` (or any other array of strings) with:
+
+    var args = parseArgs(spec, ARGV)
+
+The returned `args` is a `Dict` containing key-value pairs with the parsed
+values (or defaults) for each flag and argument. For example, given
+`ARGV = :| mysrc -P 12 mydest a b c |`, `args` would be:
+
+    {
+        "verbose": false,
+        "max-procs": 12,
+        "invert": true,
+        "src": "mysrc",
+        "dest": "mydest",
+        "files": ["a", "b", "c"]
+    }
+
+### parser
+
+`parseArgs()` requires a parser specification to indicate how to parse the
+`ARGV` array. This specification should be constructed using the `parser` proc.
+
+    parser (&spec) {
+      flag -f --my-flag
+      arg myarg
+      rest otherArgs
+    }
+
+In the above example, `parser` takes in a place `&spec`, which will store the
+resulting specification and a block which is evaluated to build that
+specification.
+
+Inside of a `parser` block, you should call the following procs:
+
+- `flag` to add `--flag` options
+- `arg` to add positional arguments
+- `rest` to capture remaining positional arguments into a list
+
+`parser` will validate the parser specification for errors such as duplicate
+flag or argument names.
+
+    parser (&spec) {
+      flag -n --name
+      flag -n --name  # Duplicate!
+    }
+
+    # => raises "Duplicate flag/arg name 'name' in spec" (status = 3)
+
+### flag
+
+`flag` should be called within a `parser` block.
+
+    parser (&spec) {
+      flag -v --verbose
+    }
+
+The above example declares a flag "--verbose" and a short alias "-v".
+`parseArgs()` will then store a boolean value under `args.verbose`:
+- `true` if the flag was passed at least once
+- `false` otherwise
+
+Flags can also accept values. For example, if you wanted to accept an integer count:
+
+    parser (&spec) {
+      flag -N --count ('int')
+    }
+
+Calling `parseArgs` with `ARGV = :| -n 5 |` or `ARGV = :| --count 5 |` will
+store the integer `5` under `args.count`. If the user passes in a non-integer
+value like `ARGV = :| --count abc |`, `parseArgs` will raise an error.
+
+Default values for an argument can be set with the `default` named argument.
+
+    parser (&spec) {
+      flag -N --count ('int', default=2)
+
+      # Boolean flags can be given default values too
+      flag -O --optimize ('bool', default=true)
```

Comment:
How do you set it to false?  (may have missed it on the thread)

Historically there are two solutions I think:

- GNU is `--no-optimize`
- Go is `--optimize=0` , but NOT `--optimize 0` as that's 2 words

That's the first thing that pops into my mind when it says `--verbose` is true if it's passed

----

OK now I see the TODO below!  I think we should follow some existing example like GNU, see if they have a special API for `--no-optimize` or `--optimize=0`